### PR TITLE
Store User Input answers in DB

### PR DESCRIPTION
**Pull request by nfmohit**
## Summary

Addresses issue:

- #5898 

## Relevant technical choices

This PR updates the User Input back-end infrastructure so that the answers are saved to the site database instead of the proxy.

The implementation totally deviates from the IB because it didn't consider saving the `site` scoped settings in the `options` table. This PR removes the old `User_Input_Settings` class and adds a new one with accompanying classes for saving respective settings in `options` and `usermeta` tables.

### Note

With this PR, the User Input answers will be wiped out if a user signs into Site Kit. This should be resolved once #5900 is merged. See discussion [here](https://github.com/google/site-kit-wp/pull/6155#discussion_r1033482404).

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/User_Input/User_Input_Settings.php [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Context        $context         Plugin context.
+	 * @param Authentication $authentication  Optional. Authentication instance. Default a new instance.
+	 */
+	public function __construct( Context $context, Authentication $authentication = null ) {
+		$options      = new Options( $context );
+		$user_options = new User_Options( $context );
```
Comment (kuasha420) [inline includes/Core/User_Input/User_Input_Settings.php:82]: Let's pass these to the constructor where `User_Input_Settings` is instantiated instead of instantiating them here. I think we can do that in the `Authentication` class where we're registering `User_Input`.

File: includes/Core/User_Input/User_Input_User_Settings.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\User_Input\User_Input_User_Settings
```
Comment (kuasha420) [inline includes/Core/User_Input/User_Input_User_Settings.php:3]: Same as above. Does `User_Specific_Questions` make sense? Let me know!

File: includes/Core/REST_API/REST_Routes.php [hunk 3]
```diff
@@ -14,7 +14,7 @@
 use Google\Site_Kit\Core\Modules\Modules;
 use Google\Site_Kit\Core\Permissions\Permissions;
 use Google\Site_Kit\Core\Authentication\Authentication;
-use Google\Site_Kit\Core\Util\User_Input_Settings;
+use Google\Site_Kit\Core\User_Input\User_Input_Settings;
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/REST_API/REST_Routes.php:17]: Done, thank you!

File: includes/Core/REST_API/REST_Routes.php [hunk 4]
```diff
@@ -193,19 +203,12 @@ private function get_routes() {
...
-											'purpose' => array(
-												'type'  => 'array',
-												'items' => array( 'type' => 'string' ),
-											),
-											'postFrequency' => array(
-												'type'  => 'array',
-												'items' => array( 'type' => 'string' ),
-											),
-											'goals'   => array(
+										'properties' => array_fill_keys(
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/REST_API/REST_Routes.php:206]: Updated, thank you!

File: includes/Core/User_Input/User_Input_Settings.php [hunk 5]
```diff
@@ -0,0 +1,313 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\User_Input\User_Input_Settings
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input_Settings.php:3]: Done, thank you!

File: includes/Core/User_Input/User_Input_Settings.php [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+ * @copyright 2022 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\User_Input;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\Authentication\User_Input_State;
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input_Settings.php:15]: As discussed in Slack, ignoring this since `User_Input_State` will be removed as a part of [#5900](https://github.com/google/site-kit-wp/issues/5900). Thank you!

File: includes/Core/User_Input/User_Input_Settings.php [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+	 */
+	protected $user_input_user_settings;
+
+	/**
+	 * User Input properties.
+	 *
+	 * @since n.e.x.t
+	 * @var array|ArrayAccess
+	 */
+	private static $properties = array(
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input_Settings.php:60]: Done, thank you!

File: includes/Core/User_Input/User_Input_Settings.php (hunk 1 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input_Settings.php:82]: Done, thank you!

File: includes/Core/User_Input/User_Input_Settings.php [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+	}
+
+	/**
+	 * Gets user input settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array|WP_Error User input settings.
+	 */
+	public function get_settings() {
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input_Settings.php:117]: Updated, thank you!

File: includes/Core/User_Input/User_Input_Settings.php [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+	/**
+	 * Sets user input settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $settings User settings.
+	 * @return array|WP_Error User input settings.
+	 */
+	public function set_settings( $settings ) {
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input_Settings.php:204]: Updated, thank you!

File: includes/Core/User_Input/User_Input_User_Settings.php (hunk 2 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input_User_Settings.php:3]: Updated, thank you!

**Comment by nfmohit** [comment]
Thank you for the kind review, @kuasha420!

I have updated the PR addressing your comments.

File: includes/Core/Authentication/Authentication.php [hunk 10]
```diff
@@ -1253,7 +1254,7 @@ private function require_user_input() {
 
 		// Refresh user input settings from the proxy.
 		// This will ensure the user input state is updated as well.
-		$this->user_input_settings->set_settings( null );
+		$this->user_input->set_answers( null );
```
Comment (kuasha420) [inline includes/Core/Authentication/Authentication.php:1257]: I know we have a separate issue to tackle this, but I was wondering, does this wipe out the settings answers in the database on every refresh at the moment? If so, may be worth having a note about it somewhere.

File: includes/Core/User_Input/Site_Specific_Questions.php [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	/**
+	 * Gets the callback for sanitizing the setting's value before saving.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Callback method that filters or type casts invalid setting values.
+	 */
+	protected function get_sanitize_callback() {
+		return function ( $settings ) {
+			return User_Input::sanitize_settings( $this, $settings, 'site' );
```
Comment (kuasha420) [inline includes/Core/User_Input/Site_Specific_Questions.php:60]: It made sense to keep the `sanitize_settings` in the `User_Input` class when it was called `User_Input_Settings`. but I think it would be better to keep the `sanitized_settings` to the corresponding `**_Specific_Settings` class.

File: includes/Core/User_Input/User_Input.php [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+			),
+		);
+	}
+
+	/**
+	 * Gets user input settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array|WP_Error User input settings.
```
Comment (kuasha420) [inline includes/Core/User_Input/User_Input.php:186]: The PHPDoc seems not up to date with the function name change. :)

File: includes/Core/User_Input/User_Input.php [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+	 */
+	public function get_answers() {
+		$data = array(
+			'site' => $this->site_specific_questions->get(),
+			'user' => $this->user_specific_questions->get(),
+		);
+
+		$questions = static::$questions;
+		$settings  = array_merge( $data['site'], $data['user'] );
```
Comment (kuasha420) [inline includes/Core/User_Input/User_Input.php:195]: Can't we use `array_merge` directly as we're not using the `$data` variable after deriving `$settings` from it?

File: includes/Core/User_Input/User_Input.php [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+
+	/**
+	 * Sets user input settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $settings User settings.
+	 * @return array|WP_Error User input settings.
+	 */
+	public function set_answers( $settings ) {
```
Comment (kuasha420) [inline includes/Core/User_Input/User_Input.php:275]: We need to update the PHPDoc here as well.

File: includes/Core/User_Input/User_Input.php [hunk 15]
```diff
@@ -0,0 +1,384 @@
...
+	 * Method re-usable by scoped classes to sanitize settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Site_Specific_Questions|User_Specific_Questions $setting  Instance of either setting class.
+	 * @param array                                           $settings Settings to sanitize.
+	 * @param string                                          $scope    Scope of the settings, either 'site' or 'user'.
+	 * @return array Sanitized settings.
+	 */
+	public static function sanitize_settings( $setting, $settings, $scope ) {
```
Comment (kuasha420) [inline includes/Core/User_Input/User_Input.php:316]: See my above suggestion above. Do you think it's possible to make this code simpler to follow somehow? I don't have any specific suggestion, but it feels hard to follow to me. :slightly_smiling_face:

File: includes/Core/User_Input/User_Specific_Questions.php [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+	/**
+	 * Gets the callback for sanitizing the setting's value before saving.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return Closure
+	 */
+	protected function get_sanitize_callback() {
+		return function ( $settings ) {
+			return User_Input::sanitize_settings( $this, $settings, 'user' );
```
Comment (kuasha420) [inline includes/Core/User_Input/User_Specific_Questions.php:61]: See my above suggestions regarding the `sanitize_settings`. Applicable here as well.

**Review by kuasha420 — CHANGES_REQUESTED** [review]
Thanks @nfmohit for the follow up commits and taking care of my suggestions. :muscle: I have left a few more suggestions for you. Let me know what you think.

File: includes/Core/User_Input/User_Input.php (hunk 12 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated, thank you!

File: includes/Core/User_Input/Site_Specific_Questions.php (hunk 11 above)
Comment (nfmohit) [reply includes/Core/User_Input/Site_Specific_Questions.php]: Done, thank you!

File: includes/Core/User_Input/User_Input.php (hunk 13 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated, thank you!

File: includes/Core/User_Input/User_Input.php (hunk 14 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated, thank you!

File: includes/Core/Authentication/Authentication.php (hunk 10 above)
Comment (nfmohit) [reply includes/Core/Authentication/Authentication.php:1257]: It doesn't wipe out on every refresh, but it does get wiped out on the `googlesitekit_authorize_user` hook, i.e. once a user signs into Site Kit. It should be removed as a part of #5900. I have added a note for it in the QA Brief and it the PR description.

**Review by kuasha420 — APPROVED** [review]
Thanks @nfmohit ! This looks good to me. I'll move it to MR and assign it to @aaemnnosttv as par his request. Cheers.

File: includes/Core/User_Input/User_Input.php [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+			$answered_by = intval( $setting['answeredBy'] );
+			unset( $setting['answeredBy'] );
+
+			if ( ! $answered_by || $answered_by === $user_id ) {
+				continue;
+			}
+
+			$setting['author'] = array(
+				'photo' => get_avatar_url( $answered_by ),
+				'name'  => get_the_author_meta( 'user_email', $answered_by ),
```
Comment (aaemnnosttv) [inline includes/Core/User_Input/User_Input.php:234]: `get_the_author_meta` is more for front-end use I believe so I don't think that's needed here. Also, the key here is `name` but this is providing their email. Does it need to be `name` or could it be `login` instead like we have elsewhere?

https://github.com/google/site-kit-wp/blob/a1327c4f3c572a326d51fc6fe5e7c695e2859f98/includes/Core/Modules/Modules.php#L1405

Looking at the other similar case we have, I'm surprised to see it also uses the same function 😄 IMO this could be refactored with the following

```php
( new WP_User( $answered_by ) )->user_email
```

[The design shows what looks like a display name](https://www.figma.com/file/CdtrHB0HAjfGRWTnTMMQI0/SummaryWidget?node-id=2074%3A61302&t=gWL4H68yse0QECLp-4) but we should use their username for consistency as this is how we reference other users in other places like notices/errors/DS, etc.

File: includes/Core/User_Input/User_Input.php [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+	/**
+	 * Sets user input answers.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array $settings User settings.
+	 * @return array|WP_Error User input answers.
+	 */
+	public function set_answers( $settings ) {
+		$questions     = static::$questions;
```
Comment (aaemnnosttv) [inline includes/Core/User_Input/User_Input.php:285]: Nit: there isn't much benefit in referencing this as a separate variable if it isn't being mutated, so this could be inlined.

File: includes/Core/User_Input/User_Specific_Questions.php [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+				}
+
+				$valid_values          = array();
+				$valid_values['scope'] = $setting_values['scope'];
+
+				$valid_answers = array();
+
+				// Verify that each answer value is a string.
+				foreach ( $setting_values['values'] as $answer ) {
+					if ( is_string( $answer ) ) {
```
Comment (aaemnnosttv) [inline includes/Core/User_Input/User_Specific_Questions.php:97]: Why specifically must the answers be strings? Maybe `is_scalar` would be a more appropriate condition? I suppose they're all strings today, but it might be odd if we chose to accept a number in the future and it were to filter that out because of this strict condition. Just a thought. What do you think?

File: tests/phpunit/integration/Core/User_Input/User_InputTest.php [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+			'values' => array(),
+		),
+	);
+
+	/**
+	 * Settings with dummy values.
+	 *
+	 * @var array|ArrayAccess
+	 */
+	private static $dummy_settings = array(
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Core/User_Input/User_InputTest.php:67]: Some of the language here goes against Google's standards. See https://github.com/google/site-kit-wp/issues/3031

Additionally, are these expected to be valid or invalid? Perhaps something like `valid_non_empty_settings` would be better?

File: tests/phpunit/integration/Core/User_Input/User_InputTest.php [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+				'postFrequency' => array( 'daily' ),
+				'goals'         => array( 'goal1', 'goal2' ),
+			)
+		);
+
+		$this->assertEquals(
+			static::$dummy_settings,
+			$response
+		);
+	}
```
Comment (aaemnnosttv) [inline tests/phpunit/integration/Core/User_Input/User_InputTest.php:199]: These are a bit confusing use of the property as there is no clear relationship between what happens in the test and the expected outcome that the value should equal the dummy settings. When using a reference from a property like this, it helps as a reusable value when we aren't overly concerned about what it is. In this case, the test is explicit about the input and ambiguous about the expectation. Another way to think about this, is that as an arbitrary value of valid settings, I would expect to be able to modify that property in any way so long that the values are valid without breaking any of the tests that use it.

File: includes/Core/User_Input/User_Input.php [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+
+		$this->site_specific_questions->set( $site_settings );
+		$this->user_specific_questions->set( $user_settings );
+
+		$updated_settings = $this->get_answers();
+		$is_empty         = $this->are_settings_empty( $updated_settings );
+
+		if ( ! is_null( $is_empty ) ) {
+			$this->authentication->get_user_input_state()->set( $is_empty ? User_Input_State::VALUE_MISSING : User_Input_State::VALUE_COMPLETED );
+		}
```
Comment (aaemnnosttv) [inline includes/Core/User_Input/User_Input.php:310]: We probably shouldn't be doing anything with `User_Input_State` since it is going to be removed as part of this epic. This also seems to be the only reason for this class as a dependency. If we need to, let's wait on #5900 before resuming work on this.

File: includes/Core/User_Input/User_Input.php [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+				$questions,
+				function ( &$question ) {
+					$question['values'] = array();
+				}
+			);
+
+			return $questions;
+		}
+
+		$user_id = get_current_user_id();
```
Comment (aaemnnosttv) [inline includes/Core/User_Input/User_Input.php:218]: It might be better to source this from `user_options->get_user_id()` instead since that is essentially the "effective user" this class is referring to.

File: includes/Core/User_Input/User_Input.php [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+	 * @since n.e.x.t
+	 *
+	 * @param Context        $context         Plugin context.
+	 * @param Authentication $authentication  Optional. Authentication instance. Default a new instance.
+	 * @param Options        $options         Optional. Options instance. Default a new instance.
+	 * @param User_Options   $user_options    Optional. User_Options instance. Default a new instance.
+	 */
+	public function __construct(
+		Context $context,
+		Authentication $authentication = null,
```
Comment (aaemnnosttv) [inline includes/Core/User_Input/User_Input.php:87]: We probably shouldn't need `Authentication here`. See below.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @nfmohit, this is off to a great start. I've left a few comments below where there are some things left to address. Nice work though 👍

**Comment by nfmohit** [comment]
Thank you so much for the very detailed and thorough review, @aaemnnosttv! I'll get boots on this again soon and get back with addressing the kind feedback. Thank you!

File: includes/Core/User_Input/User_Input.php (hunk 22 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php:310]: Understood, thank you, @aaemnnosttv! The only reason we're doing this because without it, the `User_Input_State` isn't set to a `completed` state, hence the user is stuck in the user-input screen without being able to navigate to the dashboard.

Also, #5900 is again blocked by #6155 (the issue this PR addresses). How do you think we should proceed regarding this? Thanks!

File: includes/Core/User_Input/User_Input.php [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+	 */
+	protected $site_specific_questions;
+
+	/**
+	 * User_Specific_Questions instance.
+	 *
+	 * @since n.e.x.t
+	 * @var User_Specific_Questions
+	 */
+	protected $user_specific_questions;
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated, thank you!

File: includes/Core/User_Input/User_Input.php [hunk 26]
```diff
@@ -0,0 +1,314 @@
...
+	public function register() {
+		$this->site_specific_questions->register();
+		$this->user_specific_questions->register();
+
+		add_filter(
+			'googlesitekit_rest_routes',
+			function( $routes ) {
+				return array_merge( $routes, $this->get_rest_routes() );
+			}
+		);
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Done, very good call, thank you!

File: includes/Core/User_Input/User_Input.php [hunk 27]
```diff
@@ -0,0 +1,314 @@
...
+	}
+
+	/**
+	 * Gets related REST routes.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array List of REST_Route objects.
+	 */
+	private function get_rest_routes() {
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Addressed.

File: includes/Core/User_Input/User_Input.php (hunk 17 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated to use `login` as key and `( new WP_User( $answered_by ) )->user_login` to show username as instructed. Thanks!

File: includes/Core/User_Input/User_Input.php [hunk 28]
```diff
@@ -0,0 +1,314 @@
...
+		}
+
+		$empty_settings = array_filter(
+			$settings,
+			function( $setting ) {
+				return empty( $setting['values'] );
+			}
+		);
+
+		return 0 < count( $empty_settings );
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated, thank you!

File: includes/Core/User_Input/User_Input.php (hunk 18 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated, thank you!

File: includes/Core/User_Input/User_Specific_Questions.php (hunk 19 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Specific_Questions.php]: Valid call. Updated to use `is_scalar()`, thank you!

File: tests/phpunit/integration/Core/User_Input/User_InputTest.php (hunk 21 above)
Comment (nfmohit) [reply tests/phpunit/integration/Core/User_Input/User_InputTest.php]: Removed the properties entirely and inlined the expectations, to be more explicit and readable. Please let me know it this looks good now, thank you!

File: includes/Core/User_Input/User_Input.php (hunk 23 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated, thanks!

**Comment by nfmohit** [comment]
I have addressed your valuable feedback, @aaemnnosttv!

I have made changes addressing all of them, except [this one](https://github.com/google/site-kit-wp/pull/6155#discussion_r1038462398) where I have a follow-up question.

Please let me know if further iterations are needed. Thank you!

_**Note for future reference:** The failed E2E tests are being handled in #6277 and #6284. The failed VR tests are being handled in #6185._

File: includes/Core/Authentication/Authentication.php [hunk 29]
```diff
@@ -85,13 +85,13 @@ final class Authentication {
 	private $user_input_state = null;
 
 	/**
-	 * User_Input_Settings
+	 * User_Input
 	 *
 	 * @since 1.20.0
```
Comment (aaemnnosttv) [inline includes/Core/Authentication/Authentication.php:90]: This should probably increment here as the property is different.

File: includes/Core/User_Input/User_Input.php (hunk 22 above)
Comment (aaemnnosttv) [reply includes/Core/User_Input/User_Input.php:247]: Hmm, as a temporary workaround I think we can add an action here, e.g. `googlesitekit_user_input_set` (passing along `$is_empty` as a param) and then hook into that from `User_Input_State::register` (we'd need to override it to do this). Then when that class is removed we can remove the action here as well. If we do that then we should make a note on that issue to make sure it gets cleaned up in the process.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @nfmohit ! This is looking really good. I've followed up on a few diffs but I think this is pretty close. I might need to look over it again because there is a fair amount here but it looks like everything is there. Nice work!

File: includes/Core/Authentication/Authentication.php (hunk 29 above)
Comment (nfmohit) [reply includes/Core/Authentication/Authentication.php]: Updated, thank you!

File: includes/Core/User_Input/User_Input.php [hunk 30]
```diff
@@ -0,0 +1,251 @@
...
+	 * @param Options        $options         Optional. Options instance. Default a new instance.
+	 * @param User_Options   $user_options    Optional. User_Options instance. Default a new instance.
+	 */
+	public function __construct(
+		Context $context,
+		Authentication $authentication = null,
+		Options $options = null,
+		User_Options $user_options = null
+	) {
+		$this->authentication        = $authentication ?: new Authentication( $context );
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Understood, will keep this in mind, thank you for the added context. Removed the `Authentication` instance entirely here, thanks to the new `googlesitekit_user_input_set` action.

File: includes/Core/User_Input/User_Input.php (hunk 24 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Removed, thank you!

File: includes/Core/User_Input/User_Input.php (hunk 22 above)
Comment (nfmohit) [reply includes/Core/User_Input/User_Input.php]: Updated to add the hook. Also added a note in #5900 [here](https://github.com/google/site-kit-wp/issues/5900). Thank you!

**Comment by nfmohit** [comment]
Thank you for the prompt feedback, @aaemnnosttv! Back to you for another pass.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @nfmohit – this looks good to go now! There were a few tiny fixes I pushed, one of which wasn't even really related but wasn't worth opening an issue for. Just waiting to see the tests then this should be G2G 👍

**Review by aaemnnosttv — APPROVED** [review]
Thanks @nfmohit !

File: tests/phpunit/integration/Core/User_Input/REST_User_Input_ControllerTest.php [hunk 31]
```diff
@@ -0,0 +1,137 @@
...
+use Google\Site_Kit\Core\REST_API\REST_Routes;
+use Google\Site_Kit\Core\User_Input\REST_User_Input_Controller;
+use Google\Site_Kit\Core\User_Input\User_Input;
+use Google\Site_Kit\Tests\RestTestTriat;
+use Google\Site_Kit\Tests\TestCase;
+use WP_REST_Request;
+
+class REST_User_Input_ControllerTest extends TestCase {
+
+	use RestTestTriat;
```
[the remark this answers is not in the document]
Comment (nfmohit) [reply tests/phpunit/integration/Core/User_Input/REST_User_Input_ControllerTest.php]: Oh no, I didn't notice that 😅 Thank you for updating!

**Comment by nfmohit** [comment]
Thank you for the updates and the kind review, @aaemnnosttv!
